compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp

```diff
@@ -33,46 +33,61 @@ bool resolveCPUAndCPUFeatures(llvm::StringRef inCpu,
                               llvm::StringRef inCpuFeatures,
                               const llvm::Triple &triple, std::string &outCpu,
                               std::string &outCpuFeatures) {
-  // Resolve "host"
-  if (inCpu == "host" || inCpuFeatures == "host") {
-    // If either Cpu or CpuFeatures is "host", the other must be either also
-    // host or the default value.
-    bool isCpuHostOrDefault =
-        inCpu.empty() || inCpu == "host" || inCpu == "generic";
-    bool isCpuFeaturesHostOrDefault =
-        inCpuFeatures.empty() || inCpuFeatures == "host";
-    if (!(isCpuHostOrDefault && isCpuFeaturesHostOrDefault)) {
-      llvm::errs()
-          << "error: If either cpu or CpuFeatures is `host`, the other must "
-             "be either also `host` or the default value\n";
+  // Note: only set out* variables right before returning if successful.
+  std::string cpu = inCpu.str();
+  std::string cpuFeatures = inCpuFeatures.str();
+
+  // Resolve "host" cpu and cpu features.
+  // Note: CPU feature detection in LLVM may be incomplete. Passing explicit
+  // feature lists instead of "host" should be preferred when possible.
+  if (cpu == "host" || cpuFeatures == "host") {
+    bool isCpuHostOrEmpty = cpu.empty() || cpu == "host";
+    bool isCpuFeaturesHostOrEmpty =
+        cpuFeatures.empty() || cpuFeatures == "host";
+    if (!(isCpuHostOrEmpty && isCpuFeaturesHostOrEmpty)) {
+      llvm::errs() << "error: If either Cpu or CpuFeatures is 'host', the "
+                      "other must also be 'host' or empty (Cpu: '"
+                   << cpu << "', CpuFeatures: '" << cpuFeatures << "')\n";
+      // Note: not populating outCpu or outCpuFeatures here! Fatal error later.
       return false;
     }
-    outCpu = triple.isX86() ? llvm::sys::getHostCPUName().str() : "";
+    cpu = llvm::sys::getHostCPUName().str();
     llvm::SubtargetFeatures features;
     for (auto &feature : llvm::sys::getHostCPUFeatures()) {
       features.AddFeature(feature.first(), feature.second);
     }
-    outCpuFeatures = features.getString();
+    cpuFeatures = features.getString();
   } else {
-    outCpu = inCpu;
-    outCpuFeatures = inCpuFeatures;
+    if (cpu.empty()) {
+      // TODO(#18561): Require explicit "generic" and treat unset as an error?
+      llvm::errs() << "warning: LLVMTarget cpu unspecified, defaulting to "
+                      "`generic`. Performance may suffer. Use "
+                      "`--iree-llvmcpu-target-cpu=` or the target attribute to "
+                      "set the target cpu.\n";
```

Comment:
Maybe this warning should only be logged if `cpuFeatures` is also empty.

Reply:
Indeed. Want to chat?

Reply:
Can link to docs if the message gets long: https://iree.dev/guides/deployment-configurations/cpu/
Source for docs: https://github.com/iree-org/iree/blob/main/docs/website/docs/guides/deployment-configurations/cpu.md

(Any logs that are dependent on the code running, like "list all features for the given CPU" would still want to be put here though)